File: java/libgsc/src/actor/ActorNet.java
package actor;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;

import misc.Log;
import misc.Misc;
import misc.Net;
import stmp.StmpDec;
import stmp.StmpNode;

public abstract class ActorNet extends Actor
{
	public enum NetProtocol
	{
		/** 未知. */
		NP_NONE,
		/** STMP. */
		NP_STMP,
		/** WebSocket. */
		NP_WEBSOCKET,
	};

	/** 连接上的协议类型. */
	public NetProtocol np = NetProtocol.NP_NONE;
	/** websocket-handshake是否完成, 仅在np == NP_WEBSOCKET时有效. */
	public boolean wshs = false;
	/** 连接是否已建立. */
	public boolean est = false;
	/** 连接建立时间. */
	public long gts = 0L;
	/** 最后一次收到消息的时间戳. */
	public long mts = 0L;
	/** 连接失去时间(延迟关闭时使用, 这是一个未来时间). */
	public long lts = 0L;
	/** 连接标识. */
	public SocketChannel sc = null;
	/** 缓冲区. */
	public ByteBuffer bb = null;

	public ActorNet(ActorType type, SocketChannel sc, ByteBuffer bb)
	{
		super(type);
		this.sc = sc;
		this.bb = bb;
	}

	/** 消息出栈, 成功递送到本端发送缓冲区时返回true. */
	public void send(byte[] by)
	{
		if (Log.isRecord())
		{
			if (this.np == NetProtocol.NP_WEBSOCKET && !this.wshs) /* 是websocket连接, 但还未完成握手流程. */
				Log.record("\n  --> PEER: %s\n%s", Net.getRemoteAddr(this.sc), Misc.printBytes(by));
			else
				Log.record("\n  --> PEER: %s\n%s", Net.getRemoteAddr(this.sc), StmpDec.print2Str(by));
		}
		if (!this.est)
			return;
		if (this.np == NetProtocol.NP_WEBSOCKET && this.wshs) /* websocket握手已完成. */
			this.sendWebSocket(by);
		else
			this.sendStmp(by);
	}

	/** 强制关闭连接(回调evnDis). */
	public final void close()
	{
		this.getGworker().removeActorNet(this);
		this.evnDis();
	}

	/** 静默关闭, 不触发evnDis. */
	protected void closeSlient()
	{
		if (!this.est)
		{
			Log.warn("connection was lost: %s", this);
			return;
		}
		this.getGworker().removeActorNet(this);
	}

	/** 连接失去事件. */
	public abstract void evnDis();

	/** 网络消息送达事件(STMP). */
	public abstract boolean evnMsg(StmpNode root);

	/** STMP报文出栈. */
	private void sendStmp(byte by[])
	{
		try
		{
			if (this.sc.write(ByteBuffer.wrap(by)) != by.length)
			{
				this.close();
				return;
			}
		} catch (IOException e)
		{
			if (Log.isTrace())
				Log.trace(Log.trace(e));
			this.close();
			return;
		}
	}

	/** websocket报文出栈. */
	private final void sendWebSocket(byte by[])
	{
		byte dat[] = null;
		int size = 0;
		if (by.length < 0x7E) /* 小帧. */
		{
			size = by.length + 2;
			dat = new byte[size];
			dat[0] = (byte) 0x82;
			dat[1] = (byte) by.length;
			System.arraycopy(by, 0, dat, 2, by.length);
		} else if (by.length < 0x10000) /* 中帧. */
		{
			size = by.length + 4;
			dat = new byte[size];
			dat[0] = (byte) 0x82;
			dat[1] = (byte) 0x7E;
			dat[2] = Net.short2byte((short) by.length)[0];
			dat[3] = Net.short2byte((short) by.length)[1];
			System.arraycopy(by, 0, dat, 4, by.length);
		} else
		{ /* 大帧. */
			size = by.length + 10;
			dat = new byte[size];
			dat[0] = (byte) 0x82;
			dat[1] = (byte) 0x7F;
			byte tmp[] = Net.int2byte(by.length);
			dat[2] = 0;
			dat[3] = 0;
			dat[4] = 0;
			dat[5] = 0;
			dat[6] = tmp[0];
			dat[7] = tmp[1];
			dat[8] = tmp[2];
			dat[9] = tmp[3];
			System.arraycopy(by, 0, dat, 10, by.length);
		}
		this.sendStmp(dat);
	}
}

File: c/libgsc/actor/ActorNet.cpp
/*
 * ActorNet.cpp
 *
 *  Created on: Jan 31, 2015
 *      Author: root
 */

#include "ActorNet.h"
#include "../net/Lx.h"
#include "../core/Cfg.h"
#include "../core/Gstat.h"

ActorNet::ActorNet(ActorType type, const char* name, uint wk) :
		Actor(type, name, wk)
{
	this->np = NP_NONE;
	this->wshs = false;
	this->est = false;
	this->cfd = 0;
	this->dlen = 0;
	this->rbuf = (uchar*) malloc(Cfg::libgsc_peer_mtu);
	this->lts = 0ULL;
}

/** 向套接字发送一段数据. */
void ActorNet::send(uchar* dat, uint len)
{
	Gstat::incv(LIBGSC_SND_BYTES, len);
	Gstat::inc(LIBGSC_SND_MSGS);
	LOG_RECORD("\n  --> PEER: %s CFD: %d\n%s", Net::sockaddr2str(&this->peer).c_str(), this->cfd, (this->np == NP_WEBSOCKET && !this->wshs /** 是websocket连接, 但还未完成握手流程. */) ? Misc::printhex2str(dat, len).c_str() : stmpdec_print2str(dat, len).c_str())
	if (!this->est)
		return;
	if (this->np == NP_WEBSOCKET && this->wshs) /** websocket握手已完成. */
		this->sendWebSocket(dat, len);
	else
		this->sendStmp(dat, len);
}

/** STMP报文出栈. */
void ActorNet::sendStmp(uchar* dat, uint len)
{
	if (::send(this->cfd, dat, len, MSG_DONTWAIT) == len)
		return;
	if (errno == EAGAIN || errno == EWOULDBLOCK)
	{
		LOG_DEBUG("TCP buffer was full, can not send anymore, we will close this peer: %s, size: %08X", Net::sockaddr2str(&this->peer).c_str(), len)
	} else
	{
		LOG_DEBUG("client socket exception, peer: %s, cfd: %d, size: %08X, errno: %d", Net::sockaddr2str(&this->peer).c_str(), this->cfd, len, errno)
	}
	Lx::getGwk()->removeActorNet(this);
	this->evnDis();
}

/** websocket报文出栈. */
void ActorNet::sendWebSocket(uchar* dat, uint len)
{
	uchar* buf = NULL;
	int size;
	if (len < 0x7E) /** 小帧. */
	{
		size = len + 2;
		buf = (uchar*) malloc(size);
		buf[0] = 0x82;
		buf[1] = len;
		memcpy(buf + 2, dat, len);
	} else if (len < 0x10000) /** 中帧. */
	{
		size = len + 4;
		buf = (uchar*) malloc(size);
		buf[0] = 0x82;
		buf[1] = 0x7E;
		ushort s = htons((ushort) len);
		memcpy(buf + 2, &s, 2);
		memcpy(buf + 4, dat, len);
	} else /** 大帧. */
	{
		size = len + 10;
		buf = (uchar*) malloc(size);
		buf[0] = 0x82;
		buf[1] = 0x7F;
		ullong s = Net::htonll(len);
		memcpy(buf + 2, &s, 8);
		memcpy(buf + 10, dat, len);
	}
	if (::send(this->cfd, buf, size, MSG_DONTWAIT) == size)
	{
		free(buf);
		return;
	}
	free(buf);
	if (errno == EAGAIN || errno == EWOULDBLOCK)
	{
		LOG_DEBUG("TCP buffer was full, can not send anymore, we will close this peer: %s, size: %08X", Net::sockaddr2str(&this->peer).c_str(), size)
	} else
	{
		LOG_DEBUG("client socket exception, peer: %s, cfd: %d, size: %08X, errno: %d", Net::sockaddr2str(&this->peer).c_str(), this->cfd, size, errno)
	}
	Lx::getGwk()->removeActorNet(this);
	this->evnDis();
}

/** 强制关闭连接. */
void ActorNet::close()
{
	if (!this->est)
		return;
	Lx::getGwk()->removeActorNet(this);
	this->evnDis();
}

ActorNet::~ActorNet()
{
	free(this->rbuf);
}

